Working branch

maximburyak opened the pull request.

sashatim125 committed:
DocumentConventions introduced (UNTESTED)
and the integration started

sashatim125 committed:
GetCppClassName introduces
DocumentConventions refactored using GetCppClassName
minor fixes

sashatim125 committed:
more tests added to FirstClassPatchTest

sashatim125 committed:
EntityIdHelper introduced ad DocumentConventions changed accordingly
NextHiLoCommand added
DocumentStore references were changed to std::shared_ptr
minor fixes and refactoring in tests

sashatim125 committed:
easy API for using user's document's id property (EntityIdHelper etc.) + tests
generate id on the client side
HiLo + tests(uncompleted)
minor fixes

sashatim125 committed:
RavenTestDriver introduced (as the new Java-like testing platform)
all relevant test were moved to the new platform
minor fixes

sashatim125 committed:
GetStatisticsOperation added
wait_for_indexing in RavenTestDriver implemented
minor fixes

sashatim125 committed:
using std::shared_ptr of InMemoryDocumentSessionOperations and DocumentSessionImpl
instead of reference to avoid "dangling" references

ayende commented on Raven.CppClient/DocumentConventions.cpp:
```diff
@@ -0,0 +1,391 @@
+		}
+		auto result = impl::Inflector::pluralize(simple_class_name);
+
+		_cached_default_type_collection_names.insert_or_assign(type, result);
```
Is this thread safe?

sashatim125 replied:
No.
It is now.

ayende commented on Raven.CppClient/DocumentInfo.h:
```diff
@@ -51,8 +52,7 @@ namespace ravendb::client::documents::session
 
-		//TODO std::string entity_type{};
-		//std::optional<std::type_info> exact_type;//to be filled by store()
+		std::optional<std::type_index> exact_type;//to be filled by store()
```
Not sure that I like this name. It isn't the actual type of the instance, isn't it? Just it's compile time type.
So we may have a `Dog` instance that was called with `store<Animal>`, no?
Then again, give how the rest of the system works, I don't think that matters, since we capture the serializers at compile time as well

sashatim125 replied:
Perhaps it should be called "stored_type", because that is that I wanted to emphasize.

ayende commented on Raven.CppClient/DocumentStore.h:
```diff
@@ -11,37 +16,49 @@ namespace ravendb::client::documents
 
 		void set_identifier(std::string identifier) override;
 
+		void close();
```
Why do we need something separate from the destructor?

sashatim125 replied:
Because `close()` might throw and we want to avoid exceptions thrown from a dtor.

ayende replied:
+1 on not throwing from dtor.
But what errors could cause the close of the document store to throw?

sashatim125 replied:
To start with, the Store on `close()` returns the unused HiLo ranges to the server ...

ayende replied:
Failing to return the unused hilo isn't an error. It is a courtesy, any errors doing that should be silently ignored.

sashatim125 replied:
`close()` is called from within the `~DocumentStore()` ignoring exceptions.

ayende replied:
Why have it in the first place in the interface, then?

sashatim125 replied:
In Java client `close()`  is part of the `Closeable` interface.
Here I left it `public` so the user can close the store manually and treat
all would-be-thrown exceptions before actually freeing the object's memory.

sashatim125 committed:
ClientConfiguration operations + tests
commands tests
minor fixes

sashatim125 committed:
MaintenanceOperationExecutor implemented to use RequestExecutor
minor refactoring

sashatim125 committed:
AbstractIndexCreationTask , IndexCreation, IndexDefinitionBuilder introduced
GetIndexStatisticsOperation added
SpatialOptionsFactory added
Index tests added
DateTimeOffset comparison operators added

sashatim125 committed:
AbstractJavaScriptIndexCreationTask added
set_index_name() added to AbstractIndexCreationTaskBase

sashatim125 commented:
The PR can be merged.
The next one would have the Queries implementation.

sashatim125 committed:
IndexTypeExtensions added
SpatialRelation added
